### PR TITLE
[WIP] Add support for node claiming

### PR DESCRIPTION
**Pull request by lgg42**
Fixes: https://github.com/netdata/helmchart/issues/79

This PR

* add node claiming command for netdata cloud after netdata container is ready

Tests:

- [x]  check templates generation, no errors (helm template .)

Signed-off-by: Luis Garnica <lgarnica@geoblink.com>

### CODE REVIEW

**Comment by lgg42** [comment]
I mentioned here https://github.com/netdata/helmchart/issues/79#issuecomment-654361607  that init container could be a starting way. But the first thing I encountered was the following error:

```
20:45 $ k logs -f netdata-child-vtg4x -c init-claim 
cli failed, assume netdata is not running and query the on-disk config
cli failed, assume netdata is not running and query the on-disk config
Token: ****************
Base URL: https://app.netdata.cloud
Id: unknown
Rooms: 
Hostname: ip-10-30-4-245
Proxy: 
Netdata user: netdata
Generating private/public key for the first time.
Generating RSA private key, 2048 bit long modulus (2 primes)
............+++++
.............................+++++
e is 65537 (0x010001)
Extracting public key from private key.
writing RSA key
Failed to claim node with the following error message:"invalid node id"
```

So I thought maybe of claiming the node after netdata was running, now I have the following error:

```
kube-system   netdata-child-cgvbj                                                1/2     PostStartHookError: command '/bin/sh -c netdata-claim.sh -token=##7vl-FSU4ptudp6NCKNoSGNhhu3IjxMfN4LZ9266F4AFyCT2PUaL_rmSPSoYKR-npY1Gd-XJnotWma1a04m62d4DwFXzHkGsfErkPLeDh4cEQB4xIm8pxMqOUrI7########## -rooms=##587fc3-19dc-4015-bdf5-6a########## -url=https://app.netdata.cloud' exited with 5: cli failed, assume netdata is not running and query the on-disk config
cli failed, assume netdata is not running and query the on-disk config
Token: ****************
Base URL: https://app.netdata.cloud
Id: ##b449ce-03a5-4782-8d0d-a971965e1f##
Rooms: ##587fc3-19dc-4015-bdf5-6a##########
Hostname: ip-10-30-4-91
Proxy: 
Netdata user: netdata
Generating private/public key for the first time.
Generating RSA private key, 2048 bit long modulus (2 primes)
.....................................+++++
................+++++
e is 65537 (0x010001)
Extracting public key from private key.
writing RSA key
uv_pipe_connect(): no such file or directory
Make sure the netdata service is running.
The claim was successful but the agent could not be notified (0)- it requires a restart to connect to the cloud.
```

Some characters omitted with "#"

I noticed in the dashboard that after the command was run It changed to two nodes, but I don't get any more info than that.
image
image

I'm still new to Netdata's Cloud :see_no_evil:

**Comment by prologic** [comment]
> if both of .Values.child.claiming.token and .Values.child.claiming.rooms are set, define an extra volumeMount for /var/lib/netdata/cloud.d similarly to

it might be easier to have a single persistent volume for `/var/lib/netdata` as it is pretty much the de-facto storage location for the Netdata daemon. Thoughts?

**Comment by knatsakis** [comment]
> > if both of .Values.child.claiming.token and .Values.child.claiming.rooms are set, define an extra volumeMount for /var/lib/netdata/cloud.d similarly to
> 
> it might be easier to have a single persistent volume for `/var/lib/netdata` as it is pretty much the de-facto storage location for the Netdata daemon. Thoughts?

Generally, I agree.

But, I don't know well enough what else is stored in `/var/lib/netdata` besides `cloud.d` and `registry` If the Agent team thinks it's ok, I am good too.

**Comment by lgg42** [comment]
> Hi @lgg42,
> 
> Thanks a lot for your contribution.
> 
> Could you have a look at the information provided in: [#79 (comment)](https://github.com/netdata/helmchart/issues/79#issuecomment-656170217) and below?
> 
> Could you ammend your PR to:
> 
> * Not use a postStart hook
> * Pass the ```-W set2 cloud global enabled true -W set2 cloud global "cloud base url" "${URL}" -W "claim -token="${TOKEN}" -rooms="${ROOMS}" -url="${URL}"" option to the entrypoint script of the docker image, but only if **both** of `.Values.child.claiming.token` and `.Values.child.claiming.rooms` are set
> * if **both** of `.Values.child.claiming.token` and `.Values.child.claiming.rooms` are set, define an extra volumeMount for `/var/lib/netdata/cloud.d` similarly to https://github.com/netdata/helmchart/blob/9626cec86ab775e084d2e6b92de36def3adbfb9c/templates/daemonset.yaml#L153
> 
> Also, it would be nice if you could add some short documentation about the `.Values.child.claiming.*` options in the README.md
> 
> Thank you!

Hey! thanks a lot for your comment, that's good news (everyone) :upside_down_face: !! I'll find time to make and test your proposed changes. Of course I'll add the doc in README.md regarding values. Normally I try to make the functionality work and then document, just in case I have to be changing constantly the later.

**Comment by lgg42** [comment]
Hey there! I'm writing in different channels for help, but I'm havinig a weird issue. It's detailed here: https://lesma.eu/doletodo

**tldr:** the set of args "`-W set2 cloud global enabled true`" works with a local `docker run` but fails in kubernetes. I've tried different arguments based on `-W` options and they do work in kubernetes.

Error is: `Unknown -W parameter 'set2 cloud global enabled true'`

File: templates/daemonset.yaml [hunk 1]
```diff
@@ -221,5 +227,9 @@ spec:
         - name: sd-shared
           emptyDir: {}
         {{- end }}
+        {{- if and .Values.child.claiming.enabled .Values.child.claiming.token .Values.child.claiming.rooms }}
+        - name: claim
+          emptyDir: {}
```
Comment (ilyam8) [inline templates/daemonset.yaml:232]: https://github.com/netdata/helmchart/issues/79#issuecomment-656190514

do we need to persist data in `/var/lib/netdata/cloud.d`?

If so `emptyDir` doesnt do.

**Comment by lgg42** [comment]
**Update:** With the help of a fellow, node claiming is (kinda) "ok" now. Examples:

First time installing the chart:

```
netdata-child-tkb9c netdata 2020-07-11 16:15:44: netdata INFO  : MAIN : Executing agent claiming command 'netdata-claim.sh'
netdata-child-tkb9c netdata 2020-07-11 16:15:45: netdata INFO  : MAIN : Waiting for claiming command to finish.
netdata-child-tkb9c netdata 2020-07-11 16:15:45: netdata INFO  : MAIN : Agent claiming command returned with code 0
netdata-child-tkb9c netdata 2020-07-11 16:15:45: netdata INFO  : MAIN : File '/var/lib/netdata/cloud.d/claimed_id' was found. Setting state to AGENT_CLAIMED.
netdata-child-tkb9c netdata 2020-07-11 16:15:45: netdata INFO  : MAIN : File '/var/lib/netdata/cloud.d/claimed_id' was found. Setting state to AGENT_CLAIMED.
netdata-child-tkb9c netdata 2020-07-11 16:15:46: netdata INFO  : ACLK_Main : Waiting for netdata to be claimed
```

What happens when pod in node is killed (recreated)
```
netdata-child-r6mzf netdata 2020-07-11 16:22:47: netdata INFO  : MAIN : Executing agent claiming command 'netdata-claim.sh'
netdata-child-r6mzf netdata 2020-07-11 16:22:47: netdata INFO  : MAIN : Waiting for claiming command to finish.
netdata-child-r6mzf netdata 2020-07-11 16:22:48: netdata INFO  : MAIN : Agent claiming command returned with code 13
netdata-child-r6mzf netdata 2020-07-11 16:22:48: netdata ERROR : MAIN : Agent failed to be claimed with the following error message:
netdata-child-r6mzf netdata 2020-07-11 16:22:48: netdata ERROR : MAIN : "already claimed" (errno 22, Invalid argument)
netdata-child-r6mzf netdata 2020-07-11 16:22:48: netdata INFO  : MAIN : File '/var/lib/netdata/cloud.d/claimed_id' was found. Setting state to AGENT_CLAIMED.
netdata-child-r6mzf netdata 2020-07-11 16:22:49: netdata INFO  : ACLK_Main : Waiting for netdata to be claimed
```

`/var/lib/netdata/cloud.d/claimed_id` remains the same through pod restarts, example:
```
bash-4.4# cat /var/lib/netdata/cloud.d/claimed_id 
03ea88a7-22b2-457e-84eb-1090586980cb
```

Still, I've encountered three problems:

1. After a pod restart (like the example first mentioned), the node becomes unreacheable
image

2. After the first pod creation, the node is reachable but it doesn't have metrics.
image
The cluster node's have a rule in their SG (for testing) opening 19999 to the world
image
Opening the nodes's public IP and netdata port renders the following page:
image
I ignore if this is the normal behavior. 

3. If helm release is deleted (`helm delete --purge netdata`), and nodes are manually removed from netdata cloud, and then the chart is installed again, the claim fails with:
```
netdata-child-8zr87 netdata 2020-07-11 16:40:17: netdata INFO  : MAIN : Executing agent claiming command 'netdata-claim.sh'
netdata-child-8zr87 netdata 2020-07-11 16:40:17: netdata INFO  : MAIN : Waiting for claiming command to finish.
netdata-child-8zr87 netdata 2020-07-11 16:40:18: netdata INFO  : MAIN : Agent claiming command returned with code 13
netdata-child-8zr87 netdata 2020-07-11 16:40:18: netdata ERROR : MAIN : Agent failed to be claimed with the following error message:
netdata-child-8zr87 netdata 2020-07-11 16:40:18: netdata ERROR : MAIN : "already claimed" (errno 22, Invalid argument)
netdata-child-8zr87 netdata 2020-07-11 16:40:18: netdata INFO  : MAIN : File '/var/lib/netdata/cloud.d/claimed_id' was found. Setting state to AGENT_CLAIMED.
netdata-child-8zr87 netdata 2020-07-11 16:40:19: netdata INFO  : ACLK_Main : Waiting for netdata to be claimed
```

File: templates/daemonset.yaml (hunk 1 above)
Comment (ilyam8) [reply templates/daemonset.yaml:232]: That is my question

https://github.com/netdata/helmchart/issues/79#issuecomment-656187574

If we dont need to persist that data and having persistent `netdata.public.unique.id` is enough then `.Values.child.persistUniqueID` is an additional requirement for claiming.

Comment (ilyam8) [reply templates/daemonset.yaml:232]: > remains the same through pod restarts

what about updates? delete/create? `emtpyDir` doesnt work, does it?

Comment (lgg42) [reply templates/daemonset.yaml:232]: Following the comment https://github.com/netdata/helmchart/pull/111#pullrequestreview-445726861 I tried to duplicate the `nodeuid` volume and volumeMount. 

Sorry for my ignorance, but if `/var/lib/netdata/cloud.d/claimed_id` stills the same after pod restarts why would there be necessity for persistence?

Comment (lgg42) [reply templates/daemonset.yaml:232]: > > remains the same through pod restarts
> 
> what about updates? delete/create? `emtpyDir` doesnt work, does it?

Sorry I used the wrong verb, I actually meant delete/create.
When the child pod running in a node is deleted and then recreated because of the DaemonSet, `/var/lib/netdata/cloud.d/claimed_id` remains the same. Should it be the same? should it be different?

Comment (ilyam8) [reply templates/daemonset.yaml:232]: ok, did you check if it works with (for instance) image update (basically delete old pod and create a new one)? Is it still same _node_ in the cloud?

Comment (ilyam8) [reply templates/daemonset.yaml:232]: > Sorry for my ignorance, but if /var/lib/netdata/cloud.d/claimed_id stills the same after pod restarts why would there be necessity for persistence?

To handle pod update. i am not sure having `netdata.public.unique.id` is enough.

Comment (lgg42) [reply templates/daemonset.yaml:232]: > ok, did you check if it works with (for instance) image update (basically delete old pod and create a new one)? Is it still same _node_ in the cloud?

Yes, I checked it, it stills the same _node_ in the cloud yet I encounter problem 1 mentioned in https://github.com/netdata/helmchart/pull/111#issuecomment-657094001

Comment (ilyam8) [reply templates/daemonset.yaml:232]: according docs

>  When a Pod is removed from a node for any reason, the data in the emptyDir is deleted forever.

`/var/lib/netdata/cloud.d/claimed_id` should be deleted on the pod recreate.

if i understand it right `emptyDir` allows to survive only container restarts on crashes.

Comment (ilyam8) [reply templates/daemonset.yaml:232]: ok, i think `.Values.child.persistUniqueID` should be another req for node claiming and should be added to the `if and ...`, because otherwise it would be a new node every time on a pod update.

Comment (ilyam8) [reply templates/daemonset.yaml:232]: >  yet I encounter problem 1 mentioned in

i recall i have similar problems some time ago, and i had to restart netdata after install otherwise claiming was stuck in `Waiting for netdata to be claimed`.

Could you try to restart container? (like ssh to the host and docker kill `nd` child container)

Comment (lgg42) [reply templates/daemonset.yaml:232]: > according docs
> 
> > When a Pod is removed from a node for any reason, the data in the emptyDir is deleted forever.
> 
> `/var/lib/netdata/cloud.d/claimed_id` should be deleted on the pod recreate.
> 
> if i understand it right `emptyDir` allows to survive only container restarts on crashes.

I understand it the same way as you. And it makes me think then why this happens when the pod is deleted/created:
```
tdata-child-bsn8r netdata 2020-07-11 17:20:51: netdata INFO  : MAIN : Agent claiming command returned with code 13
netdata-child-bsn8r netdata 2020-07-11 17:20:51: netdata ERROR : MAIN : Agent failed to be claimed with the following error message:
netdata-child-bsn8r netdata 2020-07-11 17:20:51: netdata ERROR : MAIN : "already claimed" (errno 22, Invalid argument)
netdata-child-bsn8r netdata 2020-07-11 17:20:51: netdata INFO  : MAIN : File '/var/lib/netdata/cloud.d/claimed_id' was found. Setting state to AGENT_CLAIMED.
```

Maybe I should set the full file path `/var/lib/netdata/cloud.d/claimed_id` for the volumeMount using `subPath` ? :thinking:

Comment (ilyam8) [reply templates/daemonset.yaml:232]: Well, i think it is expected, agent is already claimed indeed - it is same agent because it is same `netdata.public.unique.id`.

So claiming script returns an error (already claimed) but still it creates `claimed_id`.

Comment (ilyam8) [reply templates/daemonset.yaml:232]: >  and nodes are manually removed from netdata cloud

if i understand it right it doesnt do much, because there is no unclaiming functionality.

File: templates/daemonset.yaml [hunk 2]
```diff
@@ -87,6 +87,13 @@ spec:
         - name: {{ .Chart.Name }}
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
+          {{- if and .Values.child.claiming.enabled .Values.child.claiming.token .Values.child.claiming.rooms }}
```
Comment (ilyam8) [inline templates/daemonset.yaml:90]: as i said we need to add `.Values.child.persistUniqueID` to the `if and ...`

It makes it long and we use same condition in 3 places which makes the whole thing error-prone in addition.

Perhaps it would be better to create a variable and `if` it, like

```yml
{{ $addClaiming := and .Values.child.claiming.enabled .Values.child.claiming.token .Values.child.claiming.rooms .Values.child.persistUniqueID
...

{{- if $addClaiming }}
...

{{- if $addClaiming }}
...

{{- if $addClaiming }}
...

```

Just an idea. What do u think?

File: templates/daemonset.yaml (hunk 1 above)
Comment (lgg42) [reply templates/daemonset.yaml:232]: > > yet I encounter problem 1 mentioned in
> 
> i recall i had similar problems some time ago, and i had to restart netdata after install otherwise claiming was stuck in `Waiting for netdata to be claimed`.
> 
> Could you try to restart container? (like ssh to the host and docker kill `nd` child container)

Done. Not good results :cry: 

in node:
```
admin@ip-10-30-4-83:~$ sudo docker ps | grep netdata/netdata
2978e72849c8        netdata/netdata                                                  "/usr/sbin/run.sh"       15 minutes ago      Up 15 minutes                           k8s_netdata_netdata-parent-0_kube-system_974a931b-f338-4542-a50c-ba369f243857_0
cd50ab0db7fa        netdata/netdata                                                  "sh -c '/usr/sbin/ru…"   15 minutes ago      Up 15 minutes                           k8s_netdata_netdata-child-bsn8r_kube-system_6afb16e6-df90-43bb-b441-de9efe04d570_0
```
```
admin@ip-10-30-4-83:~$ sudo docker kill cd50ab0db7fa
cd50ab0db7fa
```

in the cluster:
```
netdata-child-bsn8r                                                 1/2     Error     0          18m
netdata-child-bsn8r                                                 1/2     Running   1          18m
netdata-child-bsn8r                                                 2/2     Running   1          18m
```

Yet the node remains unrecheable:
image

Do you want me to attach a pastebin from the pod logs?

File: README.md [hunk 3]
```diff
@@ -100,6 +100,9 @@ Parameter | Description | Default
 `child.podAnnotationAppArmor.enabled` | Whether or not to include the AppArmor security annotation | `true`
 `child.persistUniqueID` | Whether or not to persist `netdata.public.unique.id` across restarts | `true`
 `child.configs` | Manage custom child's configs | See [Configuration files](#configuration-files).
+`child.claiming.enabled` | Enable node claiming for netdata cloud | `false`
+`child.claiming.token` | Claim token | `""`
+`child.claiming.room` | Claim room ID | `""`
```
Comment (ilyam8) [inline README.md:105]: this is space separated list of room ids

File: templates/daemonset.yaml (hunk 2 above)
Comment (lgg42) [reply templates/daemonset.yaml:90]: > as i said we need to add `.Values.child.persistUniqueID` to the `if and ...`
> 
> It makes it long and we use same condition in 3 places which makes the whole thing error-prone in addition.
> 
> Perhaps it would be better to create a variable and `if` it, like
> 
> ```yaml
> {{ $addClaiming := and .Values.child.claiming.enabled .Values.child.claiming.token .Values.child.claiming.rooms .Values.child.persistUniqueID }}
> ...
> 
> {{- if $addClaiming }}
> ...
> 
> {{- if $addClaiming }}
> ...
> 
> {{- if $addClaiming }}
> ...
> ```
> 
> Just an idea. What do u think?

Love it. Count on it.

**Comment by ilyam8** [comment]
@lgg42 

> https://github.com/netdata/helmchart/issues/79#issuecomment-657466580

tl;dr

> So we should only be claiming the parent. We're currently working on getting streamed child node data to appear on the cloud, so we won't need to have the children claimed anyway.

related: https://github.com/netdata/helmchart/issues/114 (it seems _fix_ for #114 is https://github.com/netdata/helmchart/issues/114#issuecomment-655457446)

**Comment by lgg42** [comment]
> @lgg42
> 
> > [#79 (comment)](https://github.com/netdata/helmchart/issues/79#issuecomment-657466580)
> 
> tl;dr
> 
> > So we should only be claiming the parent. We're currently working on getting streamed child node data to appear on the cloud, so we won't need to have the children claimed anyway.
> 
> related: #114 (it seems _fix_ for #114 is [#114 (comment)](https://github.com/netdata/helmchart/issues/114#issuecomment-655457446))
> 
> There is a lot of noise in this PR, i think it would be better to open a new one. (if u are still willing to fix the issue).

Thanks a lot for the **tl;dr**. I'm ok with moving to a new PR, I'm declining this and starting a new one for parent claiming and continuing from there. I definitely want to help fix this! Can't wait to have our k8s clusters in netdata cloud :muscle:
